Upgrade to Go 1.17 and switch to Go modules and Github actions

speijnik opened the pull request:
This PR upgrades Go to 1.17, switches to Go modules for dependency management and Github actions for CI

speijnik committed:
Upgrade to Go modules, target Go 1.17 and switch to Github actions for CI

Signed-off-by: Stephan Peijnik-Steinwender <speijnik@anexia-it.com>

speijnik committed:
Add Go lint workflow

Signed-off-by: Stephan Peijnik-Steinwender <speijnik@anexia-it.com>

LittleFox94 commented on tools/go.mod:
```diff
@@ -0,0 +1,9 @@
+module go.anx.io/go-anxcloud/tools
```
the module has to be changed

LittleFox94 commented on Makefile:
```diff
@@ -1,52 +1,95 @@
+	go tool cover -html=coverage.out -o coverage.html
+
+.PHONY: func-test
+func-test: tools
```
I think you can strip this entire target. It's used in go-anxcloud for integration tests against the Engine API, nothing this library will ever have.

LittleFox94 commented on README.md:
```diff
@@ -3,17 +3,17 @@ go-structmapper
 and vice-versa.
 
-## Install
+Install
```
Why the different syntax? The previous one is the preferred markdown syntax

LittleFox94 commented on common_test.go:
```diff
@@ -5,7 +5,7 @@ import (
 	"net"
 
-	"github.com/anexia-it/go-structmapper"
+	structmapper "github.com/anexia-it/go-structmapper"
```
you don't need this change

LittleFox94 commented on Makefile:
```diff
@@ -1,52 +1,95 @@
+
+.PHONY: test
+test: tools
+	tools/ginkgo run -p 			\
```
using the ginkgo runner while not using ginkgo smells like a unneeded dependency, using `go test -race ./...` probably works the same for this project

LittleFox94 commented on README.md:
```diff
@@ -3,17 +3,17 @@ go-structmapper
 
 [![license](https://img.shields.io/github/license/mashape/apistatus.svg?maxAge=2592000)](https://github.com/anexia-it/go-structmapper/blob/master/LICENSE)
 [![GoDoc](https://godoc.org/gopkg.in/anexia-it/go-structmapper.v1?status.svg)](https://godoc.org/gopkg.in/anexia-it/go-structmapper.v1)
-[![Build Status](https://travis-ci.org/anexia-it/go-structmapper.svg?branch=v1)](https://travis-ci.org/anexia-it/go-structmapper)
```
GitHub actions also give you a status badge, maybe change to that instead of removing it?

LittleFox94 reviewed (changes requested):
You can also remove the `Gopkg.yaml`/`Gopkg.lock` now.

On a general note, using a `main` branch instead of `v1` would probably be good, Go does not resolve the major version via the branch name.

speijnik committed:
Remove ginkgo

Signed-off-by: Stephan Peijnik-Steinwender <speijnik@anexia-it.com>

speijnik committed:
Remove package aliasing

Signed-off-by: Stephan Peijnik-Steinwender <speijnik@anexia-it.com>

speijnik committed:
Fix markdown header style in README.md

Signed-off-by: Stephan Peijnik-Steinwender <speijnik@anexia-it.com>

speijnik committed:
Add badges of Github Actions workflow status

Signed-off-by: Stephan Peijnik-Steinwender <speijnik@anexia-it.com>

speijnik commented:
@LittleFox94 thanks for the initial review. I believe I addressed all of your comments.
As for the default branch: the idea is to branch main off of v1 once these changes have been merged, so this is planned already.

LittleFox94 commented on tools/.gitignore:
```diff
@@ -0,0 +1,4 @@
+/tools
+/ginkgo
```
can be removed

LittleFox94 replied:
can be removed

LittleFox94 commented on tools/.gitignore:
```diff
@@ -0,0 +1,4 @@
+/tools
```
can be removed, go-anxcloud specific

LittleFox94 commented on Makefile:
```diff
@@ -1,52 +1,82 @@
+	@tools/golangci-lint run --build-tags integration ./...
+
+.PHONY: docs-lint
+docs-lint: tools
```
only now realized you didn't copy the `documentation` workflow over, which uses this target. I'd advise to copy that workflow, too.

LittleFox94 reviewed (approved):
Looks good, found some minor things you can decide yourself if worth fixing or not :)

A note to the GitHub Actions workflows: I want to refactor them in go-anxcloud to be (mostly) a single workflow with multiple jobs, maybe you want to merge them into a single workflow, too. Due to the integration tests it isn't done in go-anxcloud, yet.

speijnik committed:
Add missing documentation workflow, update Makefile and tools/.gitignore

Signed-off-by: Stephan Peijnik-Steinwender <speijnik@anexia-it.com>

LittleFox94 commented on Makefile:
```diff
@@ -1,6 +1,6 @@
 GITTAG ?= $(shell git describe --tags --always)
 GITCOMMIT ?= $(shell git log -1 --pretty=format:"%H")
-GOLDFLAGS ?= -s -w -extldflags '-zrelro -znow' -X go.anx.io/go-anxcloud/pkg/client.version=$(GITTAG)
```
good catch!

speijnik commented:
@LittleFox94 I would re-request a review, but technically you already did one for my latest changes.
Thanks for the reviews!

LittleFox94 commented:
Gladly :)
